lib/Sema/CSSimplify.cpp

```diff
@@ -2866,7 +2866,7 @@ repairViaOptionalUnwrap(ConstraintSystem &cs, Type fromType, Type toType,
                         ConstraintLocatorBuilder locator) {
   fromType = fromType->getWithoutSpecifierType();
 
-  if (!fromType->getOptionalObjectType() || toType->is<TypeVariableType>())
+  if (!fromType->getOptionalObjectType() || toType->hasTypeVariable())
```

Comment:
The case that I'm trying to fix is exactly the case you describe. `matchTypes` is matching `Optional<Optional<Int>>` and `Optional<$T0>` and in here it's binding `$T0` to `Int` directly. I don't think this makes sense to do from `repairFailures` since this is essentially just collecting potential fixes to try which can become terms in a disjunction (I know that it's usually only one fix, but the fix can be in a disjunction with other conversions) - I think it's really unexpected that `repairViaOptionalUnwrap` would bind a type variable before attempting the fix. If we do want to look through optional types to match the context with the unwrapped type, why wouldn't we want to do that via potential binding like we do for other fixes that bind type variables?

Reply:
> If we do want to look through optional types to match the context with the unwrapped type, why wouldn't we want to do that via potential binding like we do for other fixes that bind type variables? Why wouldn't we want to do that for plain type variables too?

I think the challenge would be exactly the same, we'd need to know when it's acceptable to infer something with N levels of optional stripped off and it would mean that "fix" logic would have to be introduced into binding inference as well..